src/api/users/src/routes/users.js

```diff
@@ -29,112 +34,95 @@ router.get('/:id', validateId(), async (req, res, next) => {
+// post a user with supplied info, validated by the schema
 // rejected if a user already exists with that id, otherwise user created
-router.post('/', validateUser(), async (req, res, next) => {
+router.post('/:id', validateId(), validateUser(), validateEmailHash(), async (req, res, next) => {
+  const { id } = req.params;
+  const { body } = req;
+
   try {
-    const userRef = db.collection('users').doc(`${req.body.id}`);
+    const userRef = db.doc(id);
     const doc = await userRef.get();
 
     if (doc.exists) {
-      logger.debug(`User (id: ${doc.id}) was posted by ${req.ip} but already exists.`);
-      res.status(400).json({
-        msg: `User with id ${doc.id} was requested to be added, but already exists in the db.`,
-      });
+      next(createError(400, `User with id ${id} already exists.`));
     } else {
-      const user = new User(req.body);
-      await db
-        .collection('users')
-        .doc(`${req.body.id}`)
-        .set(JSON.parse(JSON.stringify(user))); // the user object must be parsed and stringified to be persisted to firestore
-      res.status(201).json({ msg: `Added user with id: ${req.body.id}` });
-      logger.debug(`User added with id: ${req.body.id}:\n${JSON.stringify(req.body)} by ${req.ip}`);
+      const user = new User(body);
+      await db.doc(id).set(user);
+      res.status(201).json({ msg: `Added user with id: ${id}` });
     }
   } catch (err) {
     next(err);
   }
 });
 
-// put (update) a user with a supplied id, validated by the celebrateSchema
+// put (update) a user with a supplied id, validated by the schema
 // rejected if a user could not be found, otherwise user updated
-router.put('/:id', validateUser(), async (req, res, next) => {
+router.put('/:id', validateId(), validateUser(), validateEmailHash(), async (req, res, next) => {
+  const { id } = req.params;
+  const { body } = req;
+
   try {
-    const userRef = db.collection('users').doc(`${req.body.id}`);
+    const userRef = db.doc(id);
     const doc = await userRef.get();
 
     if (!doc.exists) {
-      logger.debug(`User data (id: ${doc.id}) was requested by ${req.ip} but could not be found.`);
-      res.status(404).json({
-        msg: `User with id ${doc.id} was requested to be updated, but does not exist in the db.`,
-      });
+      next(createError(404, `user ${id} not found.`));
     } else {
-      const user = new User(req.body);
-      await db
-        .collection('users')
-        .doc(`${req.body.id}`)
-        .update(JSON.parse(JSON.stringify(user))); // the user object must be parsed and stringified to be persisted to firestore
-      res.status(200).json({ msg: `Updated user ${req.body.id}` });
-      logger.debug(`User added with id: ${req.body.id}:\n${JSON.stringify(req.body)} by ${req.ip}`);
+      const user = new User(body);
+      // NOTE: doc().update() doesn't use the converter, we have to make a plain object.
+      await db.doc(id).update(user.toJSON());
```

Comment:
I think something I forgot to add initially was to check the values of the user we're trying to create vs the values specified in the body. If they're the same let's not waste another hit to update and instead respond with.. idk, `204` and then not run `db.doc(id).update(user.toJSON())` ? If the values are different then `200`.

I just tried to spam the same `put` request over and over again and they all went through.

Reply:
This feels like a premature optimization that could potentially lead to us hiding some bug we don't intend.  Let's start out by doing the obvious thing, and adjust if it becomes a problem.